man/pathexec.3
.TH qlibs: pathexec 3
.SH NAME
pathexec \- run a program within a given environment
.SH SYNTAX
.B #include \(dqpathexec.h\(dq

\fBpathexec_run\fP(const char *\fIp\fR,char **\fIa\fR,char **\fIe\fR);
.br
\fBpathexec\fP(char *const *\fIa\fR);
.br
\fBpathexec_env\fP(const char *\fIs\fR,char *\fIt\fR);
.SH DESCRIPTION
.B pathexec_run 
searches for a program named 
.IR p . 
It replaces the current process with a copy of that program. 
The main function in that program will be given arguments 
.I a 
and environment 
.IR  e .
.B pathexec_run 
looks for 
.I p 
as specified by the 
.I $PATH 
environment variable. 
.I $PATH 
is a colon-separated list of directories 
.IR d ;
.B pathexec_run 
tries 
.B execve 
on files named 
.IR d/p , 
in the order that the directories appear inside 
.IR $PATH . 
An empty directory name is treated as a single dot.

If 
.I $PATH 
is not set, 
.B pathexec_run 
uses the path 
.IR  /bin:/usr/bin ; 
i.e., it tries 
.B execve 
on 
.IR /bin/p , 
then 
.IR /usr/bin/p .

If 
.I p 
contains a slash, 
.B pathexec_run 
ignores 
.I $PATH 
and simply runs 
.B execve 
on a file named 
.IR p .

.B pathexec 
calls 
.B pathexec_run 
with program name 
.IR a[0] , 
arguments 
.IR a , 
and the same environment as the current process, 
modified as described below. 
.B pathexec 
has the same return behavior as 
.BR pathexec_run .
.B pathexec_env 
modifies the environment used by 
.BR pathexec . 
It removes a variable named 
.IR s, 
if one exists. It then adds a variable named 
.I s 
with value 
.IR t, 
if the pointer 
.I t 
is nonzero. The name 
.I s 
must not contain 
.IR = .
.SH "RETURN CODES"
Normally
.B pathexec_run
does not return, because the process has been replaced.
However, if all the
.B execve
attempts fail,
.B pathexec_run
returns, setting
.I errno
to the most interesting error returned by
.BR execve .
Furthermore,
.B pathexec_run
returns immediately if an
.B execve
attempt fails with an error other than
.IR error_noent ,
.IR error_acces ,
.IR error_perm ,
or
.IR  error_isdir .
This list is subject to change.

Normally 
.B pathexec_env 
returns 
.IR 1 . 
If it is unable to allocate memory, it returns 
.IR 0 , 
leaving the 
.B pathexec 
environment alone.
.SH "SEE ALSO"
error(3)


CHANGELOG
fehQlibs-09:	First public release based on Qlibs as 
		a common project of Kai Peter and Erwin Hoffmann.
fehQlibs-10:	Added IPV6_V6ONLY setsockoptions call.
		Improved poll handling in iopause.c considering poll RC.
		Added building of shared libraries + updated docs.
fehQlibs-11:	Added ia4_fmt, ia6_fmt and dns_cname (for *qmail).
		Changed dns_transmit lookup time constants. 
fehQlibs-12:	dns_ip, dns_cname, dns_txt, dns_name and dns_mx
		return now the number of DNS answers received 
		unlike the number of bytes for the given output. 
		Added convenience routines for forthcoming s/qmail.
fehQlibs-12a:	Fixed missing DNS A lookup in dns_ip.c.
fehQlibs-12b:	Removed (one) obsolete return value check 
		for iopause in timeoutconn.c.
fehQlibs-12c:	Checking carefully all dnsstub return codes
		while using a coherent scheme. 
fehQlibs-13:	Bugs fixed: str_copy (wrong paranthesis setting)
		UI changed:	log() -> log_who()
		Added: man str(3)
fehQlibs-13a:	Fixed wrong input boundary checking for multiple
	DNSCACHE variables settings in dns_rcip.c. (Tx. J.W.)
fehQlibs-13b:	Added missing str_copyb() function in str.c.
fehQlibs-13c:	iopause it RC < 0 forces end of connection loop.
	DNS stub return codes straightend with djbdnscurve6-36.
fehQlibs-13d:	Fixed integration bugs not displaying IPv6 addresses
		(dns_ip.c, dns_mx.c).
		Wrong dependency check in Makefile for SHAREDLIBS (tx. Alan S.). 
fehQlibs-14:	Added 'const' for most input arguments + updated man:
		byte, logmsg, str, pathexec. Patch contributed by Alan S.
		Added byte_fill and case_upper function (potentially for SRS).
		Included CLFAGS and LDFLAGS macros in Makefile.


pathexec.c
#include <unistd.h>
#include "alloc.h"
#include "error.h"
#include "stralloc.h"
#include "str.h"
#include "byte.h"
#include "env.h"
#include "pathexec.h"

/**
	@file pathexec.c
	@author djb
	@source ucspi-tcp, ucspi-ssl
	@brief populate environment after fork
*/

static stralloc plus;
static stralloc tmp;

int pathexec_env(const char *s,const char *t)
{
  if (!s) return 1;
  if (!stralloc_copys(&tmp,s)) return 0;
  if (t) {
    if (!stralloc_cats(&tmp,"=")) return 0;
    if (!stralloc_cats(&tmp,t)) return 0;
  }
  if (!stralloc_0(&tmp)) return 0;
  return stralloc_cat(&plus,&tmp);
}

int pathexec_multienv(stralloc *sa)
{
  if (!sa) return 1;
  return stralloc_cat(&plus,sa);
}

void pathexec(char * const *argv)
{
  char **e;
  unsigned int elen;
  unsigned int i;
  unsigned int j;
  unsigned int split;
  unsigned int t;

  if (!stralloc_cats(&plus,"")) return;

  elen = 0;
  for (i = 0; environ[i]; ++i)
    ++elen;
  for (i = 0; i < plus.len; ++i)
    if (!plus.s[i])
      ++elen;

  e = (char **) alloc((elen + 1) * sizeof(char *));
  if (!e) return;

  elen = 0;
  for (i = 0; environ[i]; ++i)
    e[elen++] = environ[i];

  j = 0;
  for (i = 0; i < plus.len; ++i)
    if (!plus.s[i]) {
      split = str_chr(plus.s + j,'=');
      for (t = 0; t < elen; ++t)
        if (byte_equal(plus.s + j,split,e[t]))
          if (e[t][split] == '=') {
            --elen;
            e[t] = e[elen];
            break;
          }
      if (plus.s[j + split])
        e[elen++] = plus.s + j;
      j = i + 1;
    }
  e[elen] = 0;

  pathexec_run(*argv,argv,e);
  alloc_free(e);
}

void pathexec_run(const char *file,char *const *argv,char *const *envp)
{
  char *path;
  unsigned int split;
  int savederrno;

  if (file[str_chr(file,'/')]) {
    execve(file,argv,envp);
    return;
  }

  path = env_get("PATH");
  if (!path) path = "/bin:/usr/bin";

  savederrno = 0;
  for (;;) {
    split = str_chr(path,':');
    if (!stralloc_copyb(&tmp,path,split)) return;
    if (!split)
      if (!stralloc_cats(&tmp,".")) return;
    if (!stralloc_cats(&tmp,"/"))  return;
    if (!stralloc_cats(&tmp,file)) return;
    if (!stralloc_0(&tmp)) return;

    execve(tmp.s,argv,envp);
    if (errno != ENOENT) {
      savederrno = errno;
      if ((errno != EACCES) && (errno != EPERM) && (errno != EISDIR)) return;
    }

    if (!path[split]) {
      if (savederrno) errno = savederrno;
      return;
    }
    path += split;
    path += 1;
  }
}
